### PR TITLE
Automatically Use Large PCH on Machines With >7GB per Thread

### PR DESCRIPTION
**Pull request by NickGerleman**
### Why
We have an option to use larger PCH's which can dramatically speed up build time on high end machines. It causes build instability on machines with less memory though (e.g. hosted agents, or 8c/16T 32GB systems). The larger PCH is opt-in via environment variable because of this. Not all users with powerful machines will know the variable exists, and may not be using the optimally sized PCH for their build setup. 

### What
This change enables dynamic selection of PCH size based on available system memory. This is done via adding the `RNW_FASTBUILD` preprocessor definition based on a memory-per-thread (how much memory will a cl process get) threshold.

Large PCH saw issues with previous hosted machines, which were 3.5GB per thread. New machines have 8GB per thread, and do not see issues. Enable using large PCH in a known good scenario, so that customers will see the same things we are doing in our CI.

## Testing
Have not yet validated this locally due to some local VS issues.

We have tests now for both PCH scenarios, so we should see coverage so long as we continue to test with lower end agents. Will check that this change builds in CI and does not change runtime numbers.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9048)

### CODE REVIEW

File: vnext/PropertySheets/FastBuild.targets [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    TaskName="GetMemoryClass"
+    TaskFactory="RoslynCodeTaskFactory"
+    AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll" >
+
+    <ParameterGroup>
+      <MemoryClass ParameterType="System.String" Output="true" />
+    </ParameterGroup>
+    
+    <Task>
+      <Code Type="Fragnent" Language="cs">
```
Comment (vmoroz) [inline vnext/PropertySheets/FastBuild.targets:14]: >Fragnent [](http://example.com/codeflow?start=18&length=8)

typo in the name?

File: vnext/PropertySheets/FastBuild.targets [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (vmoroz) [inline vnext/PropertySheets/FastBuild.targets:2]: >Project [](http://example.com/codeflow?start=1&length=7)

How is this target file used by projects?

Comment (NickGerleman) [reply vnext/PropertySheets/FastBuild.targets:2]: The target will add the "RNW_FASTBUILD" definition based on the amount of memory per thread available on the system.

Comment (vmoroz) [reply vnext/PropertySheets/FastBuild.targets:2]: Should we have project files or other common targets files reference this targets file?

Comment (dannyvv) [reply vnext/PropertySheets/FastBuild.targets:2]: @vmoroz: I believe that the RNW_FASTBUILD is only used by `vnext/Microsoft.ReactNative/Pch/pch.h` so it makes sense to keep this here.. If we add it to other pch files to speed up those builds then I agreen we should move it to Directory.Build.Props

**Review by dannyvv — APPROVED** [review]

**Comment by NickGerleman** [comment]
Looks like memory class is being detected and propagated correctly, but it is not being propagated to all ClCompile items...

**Comment by NickGerleman** [comment]
Oh it looks like it is adding new items instead of replacing the metadata of them

**Comment by dannyvv** [comment]
> Oh it looks like it is adding new items instead of replacing the metadata of them

Change `<ItemGroup`  to `<ItemDefinitionGroup`

File: vnext/PropertySheets/FastBuild.targets [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    BeforeTargets="ClCompile">
+
+    <GetMemoryClass>
+      <Output PropertyName="_MemoryClass" TaskParameter="MemoryClass" />
+    </GetMemoryClass>
+
+    <Message Text="Memory class: $(_MemoryClass)" />
+
+    <!-- Add RNW_FASTBUILD as preprocessor definition to existing ClCompile items -->
+    <ItemGroup Condition="'$(RNW_FASTBUILD)' == 'true' or ('$(RNW_FASTBUILD)' == '' and '$(_MemoryClass)' == 'High')">
```
Comment (dannyvv) [inline vnext/PropertySheets/FastBuild.targets:57]:
```suggestion
    <ItemDefinitionGroup Condition="'$(RNW_FASTBUILD)' == 'true' or ('$(RNW_FASTBUILD)' == '' and '$(_MemoryClass)' == 'High')">
```

File: vnext/PropertySheets/FastBuild.targets [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      <ClCompile>
+         <PreprocessorDefinitions>RNW_FASTBUILD;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      </ClCompile>
+    </ItemGroup>
```
Comment (dannyvv) [inline vnext/PropertySheets/FastBuild.targets:61]:
```suggestion
    </ItemDefinitionGroup>
 ```

**Comment by dannyvv** [comment]
@NickGerleman : I took the liberty of updating your PR already

**Comment by NickGerleman** [comment]
Followed up offline re syntax for updating metadata, and the current is correct for this case.

**Pull request merged**